subsys/bluetooth/mesh/scene_cli.c

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <sys/byteorder.h>
+#include <bluetooth/mesh/models.h>
+#include "model_utils.h"
+
+static void scene_status(struct bt_mesh_model *mod, struct bt_mesh_msg_ctx *ctx,
+			 struct net_buf_simple *buf)
+{
+	struct bt_mesh_scene_cli *cli = mod->user_data;
+	struct bt_mesh_scene_state state;
+
+	state.status = net_buf_simple_pull_u8(buf);
+	state.current = net_buf_simple_pull_le16(buf);
+	if (buf->len == 3) {
+		state.target = net_buf_simple_pull_le16(buf);
+		state.remaining_time = net_buf_simple_pull_u8(buf);
+	} else if (!buf->len) {
+		state.target = BT_MESH_SCENE_NONE;
+		state.remaining_time = 0;
+	} else {
+		/* Invalid size */
+		return;
+	}
+
+	if (model_ack_match(&cli->ack, BT_MESH_SCENE_OP_STATUS, ctx)) {
+		struct bt_mesh_scene_state *rsp = cli->ack.user_data;
+
+		*rsp = state;
+		model_ack_rx(&cli->ack);
+	}
+
+	if (cli->status) {
+		cli->status(cli, ctx, &state);
+	}
+}
+
+static void scene_reg(struct bt_mesh_model *mod, struct bt_mesh_msg_ctx *ctx,
+		      struct net_buf_simple *buf)
+{
+	struct bt_mesh_scene_cli *cli = mod->user_data;
+	struct bt_mesh_scene_reg reg;
+
+	reg.status = net_buf_simple_pull_u8(buf);
+	reg.current = net_buf_simple_pull_le16(buf);
+	reg.count = buf->len / sizeof(uint16_t);
+	reg.scenes = net_buf_simple_pull_mem(buf, buf->len);
+
+	/* Fix endianness for big endian systems. This gets optimized to nothing
+	 * on little endian systems (like the nRF5x chips).
+	 */
+	for (int i = 0; i < reg.count; i++) {
+		reg.scenes[i] = sys_le16_to_cpu(reg.scenes[i]);
+	}
+
+	if (model_ack_match(&cli->ack, BT_MESH_SCENE_OP_REGISTER_STATUS, ctx)) {
+		struct bt_mesh_scene_reg *rsp = cli->ack.user_data;
+
+		rsp->status = reg.status;
+		rsp->current = reg.current;
+		if (rsp->scenes) {
+			rsp->count = MIN(rsp->count, reg.count);
+			memcpy(&rsp->scenes, reg.scenes,
```

Comment:
I can't find a reason to make a user allocate memory for `rsp->scenes` instead of copying the pointer to `reg.scenes`. While the `reg` callback provides the full list of received scenes, the _blocking_ call will return no more than the user requested. If the reason is to prevent the data from compromising, but then the pointer to the original data will still be provided to the user in the `reg` callback. More over, the same data will be passed to other scene clients (according to the comment on line 79).

Reply:
I'm not quite sure I follow. The data from the packet may not necessarily be valid anymore by the time the blocking API call returns, depending on thread priorities, so we have to copy it while it still is.

Reply:
Ah, right, I missed that the blocking call is handled in another thread.